adview-manager/src/helpers.rs

```diff
@@ -231,6 +236,185 @@ mod test {
+        {
+            let unit = get_unit_html(
+                Some(size_1),
+                &video_unit,
+                "https://adsense.google.com",
+                "",
+                "",
+            );
+            let fragment = Html::parse_fragment(&unit);
+
+            let anchor_selector = Selector::parse("div>a").unwrap();
+            let anchor = fragment
+                .select(&anchor_selector)
+                .next()
+                .expect("There should be a second anchor");
+
+            assert_eq!("a", anchor.value().name());
+            assert_eq!(
+                Some("https://ambire.com?utm_source=AdEx+(https://adsense.google.com)"),
+                anchor.value().attr("href")
+            );
+        }
+    }
+
+    #[test]
+    fn getting_unit_html_with_events() {
+        let whitelisted_tokens = DEFAULT_TOKENS.clone();
+
+        let market_url = ApiUrl::parse("https://market.adex.network").expect("should parse");
+        let validator_1_url = ApiUrl::parse("https://tom.adex.network").expect("should parse");
+        let validator_2_url = ApiUrl::parse("https://jerry.adex.network").expect("should parse");
+        let options = Options {
+            market_url,
+            market_slot: DUMMY_IPFS[0],
+            publisher_addr: *PUBLISHER,
+            // All passed tokens must be of the same price and decimals, so that the amounts can be accurately compared
+            whitelisted_tokens,
+            size: Some(Size::new(300, 100)),
+            navigator_language: Some("bg".into()),
+            disabled_video: false,
+            disabled_sticky: false,
+            validators: vec![validator_1_url, validator_2_url],
+        };
+        let ad_unit = AdUnit {
+            ipfs: DUMMY_IPFS[0],
+            media_url: "".into(),
+            media_mime: "video/avi".into(),
+            target_url: "https://ambire.com?utm_source=adex_PUBHOSTNAME".into(),
+        };
+        let campaign_id = DUMMY_CAMPAIGN.id;
+        let validators = DUMMY_CAMPAIGN.validators.clone();
+        // Test with events
+        {
+            let unit_with_events = get_unit_html_with_events(
+                &options,
+                &ad_unit,
+                "https://adex.network",
+                campaign_id,
+                &validators,
+                false,
+            );
+            let fragment = Html::parse_fragment(&unit_with_events);
+
+            let anchor_selector = Selector::parse("div>a").unwrap();
+            let anchor = fragment
+                .select(&anchor_selector)
+                .next()
+                .expect("There should be a second anchor");
+
+            let video_selector = Selector::parse("div>a>video").unwrap();
+            let video = fragment
+                .select(&video_selector)
+                .next()
+                .expect("There should be a video");
+
+            // TODO: If campaign.validators doesn't guarantee order this might fail and become untestable
```

Comment:
This is correct, `campaign.validators` does **not** guarantees the order will be `leader, follower`!